### DISCUSSION TITLE
Degree of self-loops (inconsistent documentation?)

### DISCUSSION THREAD

**Discussion by Splines** [networkx/networkx#6850] [Q&A] [answered]
On the [`Graph.degree`](https://networkx.org/documentation/stable/reference/classes/generated/networkx.Graph.degree.html#graph-degree) documentation page, you write:

> The node degree is the number of edges adjacent to the node. The weighted node degree is the sum of the edge weights for edges incident to that node.

But what about self-loops? Why are they counted twice?
```py
>>> import networkx as nx
>>> G = nx.Graph()
>>> G.add_edge(0,0,weight=3.14)
>>> G.degree(weight='weight')
DegreeView({0: 6.28})
```

According to your documentation, I'd expect a value of `3.14`.

Is your definition of the vertex degree inside your code common in graph theory? I'm asking since your implementation of the Louvain algorithm outputs different values compared to the original implementation by Blondel et al [here](https://sites.google.com/site/findcommunities/newversion) as they count self-loops only once for the vertex degree. If necessary I can provide an example.

**Comment by dschult** [comment, marked as answer]
There are multiple ways to handle the degree of self-loops. We double it because the outgoing and the incoming parts of the edge contribute the edge weight to the total degree. Said another way, the self-loop is incident to the node in two ways so they should both be counted. If you define your degree to only add the self-loop edge once, you lose many nice invariants -- like "the sum of the degrees over all nodes is twice the sum of the edge weights", and rewiring edges should retain the total degree of the nodes.

If you want a different definition of degree, e.g. that self-loops are only counted once in the degree, the workaround involves subtracting self-loop weights if they exist.
G_degree = {node: deg - (G[node][node][weight] if node in G[node] else 0) for node, deg in G.degree}

**Reply by Splines** [reply]
Thanks a lot for your answer, that absolutely makes sense.
